express-notes-api/index.js

```diff
@@ -0,0 +1,119 @@
+const express = require('express');
+const fs = require('fs');
+const app = express();
+const expressJson = express.json();
+
+app.use(expressJson);
+
+app.get('/api/notes', (req, res) => {
+  fs.readFile('data.json', 'utf8', (err, text) => {
+    if (err) throw err;
+    const data = JSON.parse(text);
+    const notesList = [];
+
+    for (const key in data.notes) {
+      if (key !== undefined) {
+        notesList.push(data.notes[key]);
+      } else {
+        res.json([]);
+      }
+    }
+    res.json(notesList);
+  });
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  fs.readFile('data.json', 'utf8', (err, text) => {
+    if (err) throw err;
+    const data = JSON.parse(text);
+    const id = req.params.id;
+    const dataIds = Object.keys(data.notes);
+
+    if (id > 0 && dataIds.includes(id)) {
+      res.json(data.notes[id]);
+    } else if (id > 0 && !dataIds.includes(id)) {
+      res.status(404).json({ error: `cannot find note with id ${id}` });
+    } else {
+      res.status(400).json({ error: 'id must be a positive integer' });
+    }
+  });
+});
+
+app.post('/api/notes', expressJson, (req, res) => {
+  const content = req.body.content;
+  fs.readFile('data.json', 'utf8', (err, text) => {
+    if (err) {
+      res.status(500).json({ error: 'An unexpected error occurred.' });
+    }
+    const data = JSON.parse(text);
+    if (content === undefined) {
+      res.status(400).json({ error: 'content is a required field' });
```

Comment:
In general, validation steps should be done right away if possible, rather than reading the JSON file only to find out that the client did something wrong and the file was read for nothing.